Refactor CSRF authenticator

johanjanssens opened the pull request:
This PR refactors the CSRF authenticator. 

-  The CSRF authenticator is split into a `token` and `origin` authenticator. The token authenticator uses the "Synchronizer Token Pattern" and is only activated for requests that are run in an active session context. This can be both an authentic or none-authentic user session.

- The origin authenticator is implemented according to OWASP and Stanford recommendations. If the origin or referer target don’t match the host, or are not a susdomain of the host the authentication will fail. 

- The token authenticator differentiates between AJAX and none AJAX requests, for an AJAX request the token **should** be in the custom X-CSRF-Token header, for a none AJAX request the token **should** be part of the request data.

- The  double cookie submit mechanism is removed from the token authenticator making the implementation easier to handle client side.

References:

- https://github.com/OWASP/CheatSheetSeries/blob/master/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.md#verifying-origin-with-standard-headers
- https://seclab.stanford.edu/websec/csrf/csrf.pdf

### Additional changes

- The http cookie object has been updated to allow to set the SameSite attribute, this can be used as an additional defense against csrf attacks.

- Also fixed a typo in the http cookie implementation. The http only check was happening against the wrong object parameter. see: https://github.com/joomlatools/joomlatools-framework/commit/8490f2ac867df0ea1a73bcc6c5a6468569e5a669#diff-27d571b5b08cbabbe682a37293f45b4cL203

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/authenticator/origin.php:
I would leave this one out. You don't always control all subdomain contnet. An example: User content on demo.joomlatools.com would be able to make requests to support.joomlatools.com

johanjanssens replied:
I have removed the subdomain check. Please note that this is something you added here: https://github.com/joomlatools/joomlatools-framework/issues/71

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/authenticator/csrf.php:
@johanjanssens we are not doing this anymore?

johanjanssens replied:
Removed a bit too much. We can still set the token in the header for GET requests. That way an AJAX requests can grab it more easily from the header. This approach would technically also allow us to move this into CloudFlare workers and do token validation on the edge.

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/authenticator/csrf.php:
@johanjanssens I think we used to use this to bypass CSRF checks before. Would they work now or do we need a different approach?

johanjanssens replied:
Well i considered that. A CSRF is a session hijack attempt, so technically you should protect POST requests that are running in a a session context. That means both user authentic and none authentic requests.

This approach also allows you to protect login forms with a csrf check, actually out of the box if you start a session before the authenticator runs it will kick in.

In Joomla that means you get the exact same behavior we had before, in platform if you turn sessions of it never runs unless you login

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/authenticator/token.php:
```diff
@@ -49,16 +53,21 @@ public function getCsrfToken()
-            if($request->headers->has('X-XSRF-Token')) {
-                $token = $request->headers->get('X-XSRF-Token');
-            }
+            if($request->isAjax())
```
@johanjanssens can we remove this check and accept all three cases for all requests? It's breaking form submits done through ajax in docman. (they contain the token in the data)

johanjanssens replied:
Done. All checks now run for each request

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/authenticator/origin.php:
```diff
@@ -0,0 +1,76 @@
+    {
+        parent::__construct($config);
+
+        $this->addCommandCallback('before.post', 'authenticateRequest');
```
@johanjanssens we are not running these checks on DELETE and PUT requests. Shall we not do that?

johanjanssens replied:
@ercanozkaya Not needed for DELETE and PUT as they cannot be triggered from a form, we do not to take care of method overrides as discussed.

Fix: https://github.com/joomlatools/joomlatools-framework/pull/299/commits/b0850f265a349c4cb9707cd5278a185138e4e8d7 

Please note that this makes a change in the framework by not changing the REQUEST_METHOD in the server global. Shouldn’t have impact but better double check.

ercanozkaya commented on code/libraries/joomlatools/library/dispatcher/authenticator/origin.php:
@johanjanssens Just noticed something. We do not re-define `$_SERVER['REQUEST_METHOD']`  in setMethod anymore. Since `REQUEST_METHOD` is only defined in a webserver context, this code will output a notice in CLI. We need to do an is-null check first.

johanjanssens replied:
@ercanozkaya You are right. Removed this.

ercanozkaya commented:
@johanjanssens One more problem: 
We send certain GET requests as POST requests using the method override. This allows us to circumvent the 2048 character limit. Now they are going through the CSRF token check as we use REQUEST_METHOD to check the method.

This also ties into setMethod not setting REQUEST_METHOD anymore as I reported in the other comment generating notice.

johanjanssens commented:
True. Re-wrote the code, we now check for POST, PUT, PATCH and DELETE. I also added extra hardening to getMethod() method. The override is filtered against allowed methods, if a diff method is passed get method returns null. This to prevent setting a none safe method as an override.

It does mean that by default GET through POST is not allowed. I don’t consider this a scenario we should support by default, or even at all.

Not sure how to handle the GET through POST scenario best, might be better to implement, it through a controller action override instead of a http method override?

johanjanssens commented:
@ercanozkaya Hope this are the final changes:

1. Back to checking before dispatch, this shouldn’t post any problems, especially because the other authenticators also work this way and because in Kodekit the authenticatable behavior is triggered before dispatch. 

2. I added an isPost() to the origin and token authenticators that use the raw request information to find out if it’s a post request or not.

3. Added GET to the list of allowed method overrides. Not very happy with this, ideally we would move this into Koowa and not into framework itself, from a REST perspective GET over POST isn’t ideal and should be avoided.

ercanozkaya commented:
@johanjanssens 
> I added an isPost() to the origin and token authenticators that use the raw request information to find out if it’s a post request or not.
With this change, GET over POST doesn't work. isPost will return true in the authenticator and deny the request. I'm happy to omit this from the framework but I don't really see an easy way to override it?

> Added support for defining origins. Self is added automatically to the list of origins, and the origins are used to authenticate. This will allow for CORS origin authentication
Can we add a `addOrigin` method? Otherwise you need to do getOrigin(), add, setOrigin() to add a new one.

johanjanssens commented:
### 1. GET over POST. 

GET over POST remains a POST request, you could consider this a edge-case and not to token authentication, or you could not make an exception and pass in the token, if you do that it should work. Is there any problem passing in a token for GET over POST?

### addOrigin() 

Do you really need this method? We don’t have addProxy() either. Is there a reasons that origins would need to be added at runtime? I’m trying to keep the API lean not adding methods we might need.

ercanozkaya commented:
Regarding GET on POST, yes sending a token is an option while the request is made from the site. It's the case here so I can solve it that way. 

Regarding addOrigin, for CORS you need to add them in a dispatcher behavior. By the time your behavior is called request object is already constructed. So to add a new one you need to do: 
```
$this->setOrigins(['new_one'] + $this->getOrigins());
```

which is not as clean. Also if you forget to add `+ $this->getOrigins()` to the setter the site itself won't be allowed in the origin authenticator.

johanjanssens commented:
@ercanozkaya Fair points. I refactored it further. Instead of setOrigins() I added an addOrigin() method I have also refactored getOrigin() and getReferrer() to match against the trusted origins and not just against the active request host.

_Note_: We probably don’t need the setReferrer() method. Ideally the request becomes an immutable object, something to refactor further in Kodekit.

johanjanssens commented on code/libraries/joomlatools/library/dispatcher/authenticator/token.php:
```diff
@@ -121,17 +107,20 @@ public function authenticateRequest(KDispatcherContextInterface $context)
     public function signResponse(KDispatcherContextInterface $context)
     {
-        if(!$context->response->isError())
+        if(!$context->response->isError() && $context->user->getSession()->isActive())
```
@ercanozkaya  We are using isActive() here to check. This works well in a platform context that runs without sessions, but in Joomla the session is always active. Suggest to change this to $context->user->isAuthentic() everywhere so we are only checking none guest requests.

ercanozkaya replied:
@johanjanssens Sounds good to me
